File: frontend/src/services/environmentService.ts
/**
 * Environment Service
 * Provides methods to detect and manage different environments
 */

class EnvironmentService {
  /**
   * Checks if the application is running in development mode
   * Uses Vite's import.meta.env.DEV and hostname checks
   */
  isDevelopment(): boolean {
    // Check Vite's built-in environment variable
    if (import.meta.env.DEV) {
      return true
    }

    // Also check hostname as a fallback
    const hostname = window.location.hostname.toLowerCase()
    return hostname === 'localhost' || hostname === '127.0.0.1'
  }

  /**
   * Checks if the application is running in production mode
   */
  isProduction(): boolean {
    return import.meta.env.PROD && !this.isStaging()
  }

  /**
   * Checks if the application is running in staging mode
   * Looks for "staging" in the hostname
   */
  isStaging(): boolean {
    const hostname = window.location.hostname.toLowerCase()
    return hostname.includes('staging')
  }

  /**
   * Returns the current environment name
   */
  getEnvironmentName(): 'development' | 'staging' | 'production' {
    if (this.isDevelopment()) {
      return 'development'
    }
    if (this.isStaging()) {
      return 'staging'
    }
    return 'production'
  }

  /**
   * Determines if dev login should be enabled
   * Dev login is only active in development environment
   */
  isDevLoginEnabled(): boolean {
    return this.isDevelopment()
  }
}

// Export singleton instance
export const environmentService = new EnvironmentService()
export default environmentService

File: backend/internal/services/environment_service.go
package services

import (
	"strings"

	"github.com/vibexp/vibexp/internal/config"
)

// EnvironmentService provides methods to detect and manage different environments
type EnvironmentService struct {
	config *config.Config
}

// NewEnvironmentService creates a new EnvironmentService instance
func NewEnvironmentService(cfg *config.Config) *EnvironmentService {
	return &EnvironmentService{
		config: cfg,
	}
}

// IsDevelopment checks if the application is running in development mode
// based on the FRONTEND_BASE_URL pointing at localhost or 127.0.0.1.
// Empty FRONTEND_BASE_URL is treated as NOT development (fail-closed) so
// misconfigured environments cannot accidentally enable dev-only paths. It
// delegates to config.IsLocalDevelopment, the single source of truth for the dev
// heuristic shared with the dev-only config derivation.
func (s *EnvironmentService) IsDevelopment() bool {
	return s.config.IsLocalDevelopment()
}

// IsProduction checks if the application is running in production mode
func (s *EnvironmentService) IsProduction() bool {
	return !s.IsDevelopment() && !s.IsStaging()
}

// IsStaging checks if the application is running in staging mode
// It checks if FRONTEND_BASE_URL contains "staging"
func (s *EnvironmentService) IsStaging() bool {
	url := strings.ToLower(s.config.Frontend.BaseURL)
	return strings.Contains(url, "staging")
}

// GetEnvironmentName returns the current environment name
func (s *EnvironmentService) GetEnvironmentName() string {
	if s.IsDevelopment() {
		return "development"
	}
	if s.IsStaging() {
		return "staging"
	}
	return "production"
}

// IsDevLoginEnabled determines if the /api/v1/auth/dev/login endpoint
// should respond. Both conditions must be true:
//  1. DEV_LOGIN_ENABLED env var is explicitly set to true (defaults to false).
//  2. The runtime environment is detected as development.
//
// This double-gate prevents a single env-var misconfiguration (e.g. a
// FRONTEND_BASE_URL accidentally containing "localhost") from exposing
// unauthenticated user impersonation in production.
func (s *EnvironmentService) IsDevLoginEnabled() bool {
	return s.config.Auth.DevLoginEnabled && s.IsDevelopment()
}
